libkbfs/block_tree.go

```diff
@@ -291,159 +311,214 @@ func (bt *blockTree) getBlocksForOffsetRange(ctx context.Context,
 		info, iptrOff := pblock.IndirectPtr(i)
 		// Some byte of this block is included in the left side of the
-		// range if `startOff` is less than the largest byte offset in
+		// range if `job.startOff` is less than the largest byte offset in
 		// the block.
 		inRangeLeft := true
 		if i < pblock.NumIndirectPtrs()-1 {
 			_, off := pblock.IndirectPtr(i + 1)
-			inRangeLeft = startOff.Less(off)
+			inRangeLeft = job.startOff.Less(off)
 		}
 		if !inRangeLeft {
 			continue
 		}
 		// Some byte of this block is included in the right side of
-		// the range if `endOff` is bigger than the smallest byte
+		// the range if `job.endOff` is bigger than the smallest byte
 		// offset in the block (or if we're explicitly reading all the
 		// data to the end).
-		inRangeRight := endOff == nil || iptrOff.Less(endOff)
+		inRangeRight := job.endOff == nil || iptrOff.Less(job.endOff)
 		if !inRangeRight {
 			// This block is the first one past the offset range
 			// amount the children.
-			nextBlockOffsetThisLevel = iptrOff
-			break
+			results <- result{nextBlockOffset: iptrOff}
+			return nil
 		}
 
 		childPtr := info.BlockPointer
 		childIndex := i
-		respCh := make(chan resp, 1)
-		respChans = append(respChans, respCh)
-		// Don't reference the uncaptured `i` variable below.
-		eg.Go(func() error {
-			var pfr [][]parentBlockAndChildIndex
-			var blocks map[BlockPointer]Block
-			var nextBlockOffset Offset
-			// We only need to fetch direct blocks if we've been asked
-			// to do so.  If the direct type of the pointer is
-			// unknown, we can assume all the children are direct
-			// blocks, since there weren't multiple levels of
-			// indirection before the introduction of the flag.
-			if getDirect || childPtr.DirectType == IndirectBlock {
-				block, _, err := bt.getter(
-					groupCtx, bt.kmd, childPtr, bt.file, blockReadParallel)
-				if err != nil {
-					return err
-				}
 
-				// Recurse down to the level of the child.
-				pfr, blocks, nextBlockOffset, err = bt.getBlocksForOffsetRange(
-					groupCtx, childPtr, block, startOff, endOff, prefixOk,
-					getDirect)
-				if err != nil {
-					return err
-				}
-			} else {
-				// We don't care about direct blocks, so leave the
-				// `blocks` map `nil`.
-				pfr = [][]parentBlockAndChildIndex{nil}
-				nextBlockOffset = nil
+		childPath := make([]parentBlockAndChildIndex, len(job.pathPrefix)+1)
+		copy(childPath, job.pathPrefix)
+		childPath[len(childPath)-1] = parentBlockAndChildIndex{
+			pblock:     pblock,
+			childIndex: childIndex,
+		}
+
+		// We only need to fetch direct blocks if we've been asked
+		// to do so.  If the direct type of the pointer is
+		// unknown, we can assume all the children are direct
+		// blocks, since there weren't multiple levels of
+		// indirection before the introduction of the flag.
+		if job.getDirect || childPtr.DirectType == IndirectBlock {
+			// Create a task for the child, and either enqueue it or recurse to it.
+
+			subTask := task{
```

Comment:
Maybe slightly less error-prone to do something like:
```go
subTask := job
subtask.ptr = childPtr
subtask.pblock = nil
subtask.pathPrefix = childPath
```

But I don't feel strongly about it either way.

Even better might be making a `job.subTask()` function that takes in the new data and encapsulates this copy, so the logic is closer to where the fields are defined.